enhancements/machine-api/aws-placement-groups.md

```diff
@@ -0,0 +1,485 @@
+  required capacity will be available for your launch to succeed. We recommend using the same instance type for all
+  instances in a cluster placement group.
+
+##### Spread Placement Groups
+
+A Spread placement group is a grouping of instances all placed in distinct racks within the datacenter.
+This ensures that each instance within the group has its own network and power source, maximising resillience to
+hardware failures.
+
+Spread placement groups are recommended for launching a small number of instances that must be kept separate, eg a
+cluster of etcd instances.
+
+As Spread placement groups keep instances in different racks, they can be expanded over time and can also mix instance
+types without detriment.
+
+Spread placement groups have a limited capacity. You can only launch seven instances within a Spread placement group
+within a single availability zone.
+
+###### Limitations of Spread Placement Groups
+
+- Spread placement groups are not supported for AWS Dedicated Instances.
+
+##### Partition Placement Groups
+
+A Partition placement group divides your instances into logical segments called partitions.
+Each partition is assigned a set of racks within the Availability Zone, each rack in turn having its own network and
+power source.
+No rack within the AZ will appear in more than one partition.
+
+Users may either create instances and have them spread automatically across partitions or they may allocate instances
+to a particular partition.
+Partitions are identified numerically, indexed from 1.
+
+Partition placement groups are recommended for deploying large distributed and replicated workloads across distinct
+racks.
+
+Partition placement groups are limited to seven partitions per Availability Zone.
+
+When using a Partition placement group, if users wish to control the placement of their workloads within Partitions,
+they will need to create multiple MachineSets and assign each one to a partition.
+Once they have assigned different MachineSets to partitions, they will be able to use Node affinity to assign the
+workloads to the Nodes created by the Machines within the MachineSet by selecting based on the MachineSets unique
+labels.
+
+###### Limitations of Partition Placement Groups
+
+- A partition placement group with AWS Dedicated Instances can have a maximum of two partitions.
+
+#### Creation of Placement Groups
+
+To enable users to get started with Placement Groups, we will allow users to specify the name of the placement group to
+use for the new Machines.
+If the placement group does not exist, we will create it based on the configuration provided by the user.
+
+We will allow the user to specify the type of the placement group to create, either Cluster, Spread or Partition, as
+well as, in the case of Partition, the number of partitions to be created.
+
+Should the user not define the type, we will default to Spread to ensure maximal hardware fault tolerance within the
+group.
```

Comment:
(For context for others, Scott and I had different intuition on which of these should be the default.)

Yes, I think you might be right, it's not particularly obvious if there is a sensible default for the creation of new placement groups, so I'm leaning towards - If we need to create a placement group, then the field is required

I wouldn't want to make it required in terms of API required as we don't need the field when the PG already exists.

I think there was a suggestion (I can't remember where we landed on it) of checking the placement group type matched the spec, so you would know that the placement group you are using is a cluster, spread or partition from the machine spec, and we would refuse to create a machine if the type didn't match. If we were to go down that route, then requiring the field would make sense

Reply:
> we would refuse to create a machine if the type didn't match. If we were to go down that route, then requiring the field would make sense

We could refuse to create the machine and still not require the field.  I don't think making a default here makes sense.  It's also unclear to me how you would delete decide if and when to delete a placementgroup if you auto-create.

Reply:
So I think we agreed on no default, and erroring when they don't set the field, so that's cleared.

> It's also unclear to me how you would delete decide if and when to delete a placementgroup if you auto-create.

If the group has no instances in it, and it was tagged appropriately (has the cluster ID tag), then we will delete it after we remove the last instance. The idea here is that if a user creates a PG out of band they won't tag it with the cluster ID (this may be a bit loose of an assumption), but also, they are quick to create so if a user adds a new machine and the PG doesn't exist we can create a new one easily. Perhaps having a valid configuration should be a prerequisite to deleting the PG (ie groupType is set)

Reply:
It seems like we have agreement to get rid of the default here? Maybe even that it should be required? Can we update if so.